### DISCUSSION TITLE
“NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。

### DISCUSSION THREAD

**Discussion by linzhuzhang** [dlemstra/Magick.NET#754] [Help]
ERROR:   The type initializer of "NativeMagickSettings" throws an exception. The type initializer for 'x64' threw an exception.

This is run in net framework 4.5.2. This error will be reported when calling the DLL. What is the cause? How to solve this problem?

**Comment by dlemstra** [comment]
What is the exception? Can you view the exception that was raised?

**Comment by linzhuzhang** [comment]
“An exception was thrown by the type initializer of nativemaicksettings'. #The type initializer for 'x64' threw an exception。#Magick.NET-Q16-AnyCPU#&nbsp; &nbsp;在 ImageMagick.MagickSettings.NativeMagickSettings..ctor()
&nbsp; &nbsp;在 ImageMagick.MagickSettings..ctor()
&nbsp; &nbsp;在 ImageMagick.MagickImage..ctor()
&nbsp; &nbsp;在 ZonLian.Common.DLL.ImageHelp.SaveImage(Stream vImageStream, String vImageSavePath) 位置 E:\SVN\ZonLian\developer\ZonLian.Common\ZonLian.Common.DLL\ImageHelp.cs:行号 31
&nbsp; &nbsp;&nbsp;





the&nbsp; code :


&nbsp; &nbsp; &nbsp; &nbsp; /// <summary&gt;
&nbsp; &nbsp; &nbsp; &nbsp; ///&nbsp; 保存图片
&nbsp; &nbsp; &nbsp; &nbsp; /// </summary&gt;
&nbsp; &nbsp; &nbsp; &nbsp; /// <param name="vImageStream"&gt;</param&gt;
&nbsp; &nbsp; &nbsp; &nbsp; /// <param name="vImageSavePath"&gt;</param&gt;
&nbsp; &nbsp; &nbsp; &nbsp; /// <returns&gt;</returns&gt;
&nbsp; &nbsp; &nbsp; &nbsp; public ResBase SaveImage(Stream vImageStream,string vImageSavePath)
&nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; #region&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; ResBase result = new ResBase()
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; retCode = ResCodeEnum.failed,
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; retMsg = "操作失败"
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; };
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; try
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; MagickImage image = new MagickImage(vImageStream);&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; image.Write(vImageSavePath);&nbsp;&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retCode = ResCodeEnum.success;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retMsg = "操作成功";
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; image.Dispose();
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; catch(MagickErrorException&nbsp; &nbsp;aex)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retCode = ResCodeEnum.exception;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retMsg += JsonHelp.ObjectConvertJsonString(aex);
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; catch (MagickException aex)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retCode = ResCodeEnum.exception;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result.retMsg += JsonHelp.ObjectConvertJsonString(aex);
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }&nbsp;&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; return result;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; #endregion
&nbsp; &nbsp; &nbsp; &nbsp; }





The recorded exception information is the code above and then executed






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 下午5:29
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
What is the exception? Can you view the exception that was raised?

**Comment by linzhuzhang** [comment]
It's strange that the same code configures two sites on an IIS on the same server. One site can execute normally, and the other site has an exception. Then I set the 32-bit application in the advanced settings of IIS application pool of the site with exception as true, and the abnormal site can execute normally.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 下午5:29
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
What is the exception? Can you view the exception that was raised?

**Reply by dlemstra** [reply]
You probably want to use `MagickAnyCPU.CacheDirectory` (https://github.com/dlemstra/Magick.NET/tree/27b6e2cea8a6c60af7c73c4c384ed084ee3e020c/docs#anycpu) and point to a different folder for each site. Another option would be to use `MagickAnyCPU.HasSharedCacheDirectory` but it is very likely that I will remove that from the API very soon.

**Comment by linzhuzhang** [comment]
I'll try as you say



thank&nbsp;you&nbsp; very&nbsp;much ！！！
&nbsp;&nbsp;


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 晚上6:10
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
You probably want to use MagickAnyCPU.CacheDirectory (https://github.com/dlemstra/Magick.NET/tree/27b6e2cea8a6c60af7c73c4c384ed084ee3e020c/docs#anycpu) and point to a different folder for each site. Another option would be to use MagickAnyCPU.HasSharedCacheDirectory but it is very likely that I will remove that from the API very soon.

**Comment by linzhuzhang** [comment]
Cache directory intelligent settings C disk under the directory?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 晚上6:10
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
You probably want to use MagickAnyCPU.CacheDirectory (https://github.com/dlemstra/Magick.NET/tree/27b6e2cea8a6c60af7c73c4c384ed084ee3e020c/docs#anycpu) and point to a different folder for each site. Another option would be to use MagickAnyCPU.HasSharedCacheDirectory but it is very likely that I will remove that from the API very soon.

**Comment by linzhuzhang** [comment]
I now use the root directory of the current program's execution as the cache directory, but it doesn't seem to have any effect. Its cache or go to the C:/Windows/Temp directory








------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 晚上6:10
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
You probably want to use MagickAnyCPU.CacheDirectory (https://github.com/dlemstra/Magick.NET/tree/27b6e2cea8a6c60af7c73c4c384ed084ee3e020c/docs#anycpu) and point to a different folder for each site. Another option would be to use MagickAnyCPU.HasSharedCacheDirectory but it is very likely that I will remove that from the API very soon.

**Reply by dlemstra** [reply]
Your program needs write access to that directory. I would advise you to use a different folder than the root of your program.

**Comment by linzhuzhang** [comment]
I have set the permission to write to other directories, but it still has no effect. Can you solve this problem in the next version?





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 晚上7:33
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
Your program needs write access to that directory. I would advise you to use a different folder than the root of your program.

**Reply by dlemstra** [reply]
This is not something I can solve. This is most likely an issue in your configuration. You will need to make sure that you set the `CacheDirectory` before any Magick.NET code is execute. The best place to do this would be in the startup of your application.

**Comment by linzhuzhang** [comment]
Now I find a temporary solution, which is to set the application pool of two sites to the same in "IIS". That's it. But I still want to see if it can be fundamentally solved!




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "dlemstra/Magick.NET"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月10日(星期六) 晚上7:33
收件人:&nbsp;"dlemstra/Magick.NET"<Magick.NET@noreply.github.com&gt;;
抄送:&nbsp;"曲凡"<197231519@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [dlemstra/Magick.NET] “NativeMagickSettings”的类型初始值设定项引发异常。“X64”的类型初始值设定项引发异常。 (#754)





 
Your program needs write access to that directory. I would advise you to use a different folder than the root of your program.
